File: modules/wem-api/src/main/java/com/enonic/wem/api/form/inputtype/Date.java
package com.enonic.wem.api.form.inputtype;


import com.enonic.wem.api.data.type.ValueTypes;
import com.enonic.wem.api.data2.Property;
import com.enonic.wem.api.data2.Value;
import com.enonic.wem.api.form.BreaksRequiredContractException;

final class Date
    extends InputType
{
    Date()
    {
    }

    @Override
    public void checkBreaksRequiredContract( final Property property )
        throws BreaksRequiredContractException
    {

    }

    @Override
    public Value newValue( final String value )
    {
        return Value.newLocalDate( ValueTypes.LOCAL_DATE.convert( value ) );
    }

    @Override
    public InputTypeConfig getDefaultConfig()
    {
        return null;
    }
}

File: modules/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/time/Date.ts
module api.content.form.inputtype.time {

    import support = api.form.inputtype.support;

    import Property = api.data.Property;
    import Value = api.data.Value;
    import ValueType = api.data.ValueType;
    import ValueTypes = api.data.ValueTypes;

    /**
     * Uses [[api.data.ValueType]] [[api.data.ValueTypeLocalDate]].
     */
    export class Date extends support.BaseInputTypeNotManagingAdd<any,Date> {

        constructor(config: api.form.inputtype.InputTypeViewContext<any>) {
            super(config);
        }

        getValueType(): ValueType {
            return ValueTypes.LOCAL_DATE;
        }

        newInitialValue(): Value {
            return ValueTypes.LOCAL_DATE.newNullValue();
        }

        createInputOccurrenceElement(index: number, property: Property): api.dom.Element {

            var datePickerBuilder = new api.ui.time.DatePickerBuilder();

            if (!property.hasNullValue()) {
                var date = property.getLocalDate();
                datePickerBuilder.
                    setSelectedDate(date).
                    setYear(date.getFullYear()).
                    setMonth(date.getMonth());
            }
            var datePicker = datePickerBuilder.build();
            datePicker.onSelectedDateChanged((event: api.ui.time.SelectedDateChangedEvent) => {
                var changedValue = new Value(event.getDate(), ValueTypes.LOCAL_DATE);
                property.setValue(changedValue);
            });
            return datePicker;
        }

        availableSizeChanged() {
        }

        valueBreaksRequiredContract(value: Value): boolean {
            return value.isNull() || !value.getType().equals(ValueTypes.LOCAL_DATE);
        }

        hasInputElementValidUserInput(inputElement: api.dom.Element) {
            var datePicker = <api.ui.time.DatePicker>inputElement;
            return datePicker.hasValidUserInput();
        }
    }
    api.form.inputtype.InputTypeManager.register(new api.Class("Date", Date));

}